### DISCUSSION TITLE
Can you use resonate and fire based models on loihi2 right now?

### DISCUSSION THREAD

**Discussion by Michaeljurado42** [lava-nc/lava-dl#100] [General]
Hello and good afternoon. 

Recently I tried to export a resonate and fire model (a model containing rf.Dense blocks) as a .net file so I could simulate it for loihi2 via netx. It was then that I noticed that rf.Dense blocks don't seem to be compatible with netx. I also noticed that in the tutorial section that netx is a gateway to allow models to be tested and benchmarked on loihi2. 

Are there any workarounds to demo models with rf dense layers on loihi2 without the use of netx? If not, is integrating rf.Dense layers with netx going to be a future development effort?

Thank you for any help or pointers you can provide.

**Comment by bamsumit** [comment]
Hi @Michaeljurado24 ,

Thanks for reaching out to us. __Lava-dl NetX__ does not support resonate and fire neurons right now. _The reason is that the main lava does not yet have a resonate and fire process and model available_. I am happy to guide you to contribute a resonate and fire model and enable it through netx.

**Reply by Michaeljurado42** [reply]
I've started working on a [lava fork](https://github.com/Michaeljurado24/lava/tree/rf_neurons) to enable rf neurons. For the floating implementation, I think I have implemented it correctly since it seems to match the lava-dl [output](https://user-images.githubusercontent.com/18356610/191039335-a1966b72-6bbb-4556-88de-2025ec3813a1.png). But I may soon have some questions for you about a fixed point implementation and bit accurate implementation though.

**Comment by Michaeljurado42** [comment]
So my first question regards the rf complex spiking function inside [lava-dl](https://github.com/lava-nc/lava-dl/blob/main/src/lava/lib/dl/slayer/spike/complex.py#L90). It seems that what you are doing here is essentially making sure that the neuron does not spike consecutively in two timesteps? (I am not sure about this.) I think I was able to mimic this functionality inside a floating point representation ([shown here](https://github.com/Michaeljurado24/lava/blob/rf_neurons/src/lava/proc/rf/models.py#L34)).  It seems to [work well for fixed point](https://user-images.githubusercontent.com/18356610/191564576-bb375c23-5b86-4d36-907d-2bb005799d2d.png) and match the [lava-dl implementation of the same neuron](https://github.com/lava-nc/lava-dl/files/9618871/lava-dl-result.pdf), but the same implementation for floating point leads to slightly delayed spiking as shown [in this graphic](https://user-images.githubusercontent.com/18356610/191564364-a14e44a6-6d53-42f3-98a4-4054b8201605.png).

So my question does my current interpretation of the spiking behavior look right for rf neurons and if not what is the correct spiking lava spiking behavior for this neuron? 

(I am currently using [this temporary file](https://github.com/Michaeljurado24/lava/blob/rf_neurons/test_resonator_process.py) to run small neuron tests if you want to reproduce any of the behavior)

**Reply by bamsumit** [reply]
That could happen if the fixed point implementation truncates to zero first whereas the floating-point value is still a small negative value. If that's the case, it is fine. What you have in `run_spk` seems correct at a first glance.

**Comment by Michaeljurado42** [comment]
Here is more of a design question. I am currently modifying lava-dl netx to be able to export trained rf models to callable lava processes. Do you think it is better to modify the netx.blocks.models.Dense class to be able to handle the complex case as well or that it is better to create new class called netx.blocks.models.ComplexDense  and perhaps a new method inside hdf5.py called create_complex_dense instead? Or do you think there is an easier way I am not quite seeing?

--edit--
I think that the later solution will probably be more neat so I'll go with that unless someone disagrees.

**Reply by bamsumit** [reply]
I would do the following
1. Create a dense block process and process model `netx.blocks.process.ComplexDense` explicitly or add an argument `complex_synapse=True` in the process definition. Both work for me. I will leave it up to you.
2. I would still keep `hdf5.Network.create_dense`.
    * Modify `Network.get_neuron_params` to return a bool called `complex_neuron`. It is only at this stage we can decide whether we have a complex neuron or a real neuron in the network. A complex neuron would now need a complex synapse and we can call the api in step 1 accordingly.
